Sprint 10

Fennec-meow opened the pull request.

Fennec-meow committed:
Constants.

Fennec-meow committed:
sprint_10

Fennec-meow committed:
Some adjustments have been made, as well as the "fetchOAuthToken" function has been changed

Fennec-meow committed:
Corrections:
1. The "OAuth2Service" class
Data transmission has been fixed. Now they are passed not through the "body", but through the "query" parameters.
2. The "SplashViewController" class
The "switchToTabBarController" function has been edited. Changes have been made related to the display of "TabBarViewController" in the main stream.

So1ovyov commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,56 @@
+}
+
+final class AuthViewController: UIViewController {
+    private let ShowWebViewSegueIdentifier = "ShowWebView"
```
Наименование констант, переменных и методов должно быть в стиле `CamelCase` и начинаться с нижнего регистра:
```
private let showWebViewSegueIdentifier = "ShowWebView"
```

So1ovyov commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,56 @@
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+//        configureBackButton()
```
Забыла раскомментировать метод по настройке кнопки, без этой настройки дизайн не соответствует макету.

So1ovyov commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,56 @@
+    }
+}
+    
+//    private func configureBackButton() {
```
Кажется нужно раскомментировать.

So1ovyov commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,122 @@
+import UIKit
+import WebKit
+
+fileprivate let UnsplashAuthorizeURLString = "https://unsplash.com/oauth/authorize"
```
Наименование с верхнего регистра не корректно, нужно исправить:
```
private let unsplashAuthorizeURLString = "https://unsplash.com/oauth/authorize"
```
Также предлагаю перенести константу либо в класс `WebViewViewController` либо в специальный `enum Constants`

So1ovyov commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,122 @@
+    }
+}
+
+func makeOAuthTokenRequest(code: String) -> URLRequest {
```
У тебя почему-то метод вне какой-либо сущности, это ошибка, кажется что этот метод должен быть в `OAuth2Service`

So1ovyov commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,122 @@
+        + "&&code=\(code)"
+        + "&&grant_type=authorization_code",
+        relativeTo: baseURL                           // Опираемся на основной или базовый URL, которые содержат схему и имя хоста
+    )!
```
Одно из условий ТЗ  "В проекте не используется `implicit unwrap` и другие force-операции." Так что `force` операцию нужно удалить (!)

So1ovyov commented on ImageFeed/Constants.swift:
```diff
@@ -0,0 +1,16 @@
+
+import Foundation
+
+let accessKey = "F9m5lcPo6uJuYbyngJGb_7qqdOvgPuJN8jjDSnlV8RE"
```
Условие ТЗ не соблюдено: "В проекте создан файл `Constants.swift`, в нём есть константы `accessKey`, `secretKey`, `redirectURI` и `accessScope`"

So1ovyov commented on ImageFeed/Constants.swift:
```diff
@@ -0,0 +1,16 @@
+let secretKey = "aMsHV4IoKIOr-gFXW7ErR7TDJL4NEAAf20H6DLM9ACk"
+let redirectURI = "urn:ietf:wg:oauth:2.0:oob"
+let accessScope = "public+read_user+write_likes"
+let defaultBaseURL = URL(string: "https://api.unsplash.com")!
```
Еще один `force unwrap` нужно исправить

So1ovyov commented on ImageFeed/ServiceAccessLayer/OAuth2/OAuth2Service.swift:
```diff
@@ -0,0 +1,71 @@
+            
+            request.httpMethod = "POST"
+                        
+            let task = URLSession.shared.dataTask(with: request) { data, response, error in
```
Ты создала `URLSession+data` где все что ниже у тебя в коде, уже сделано в этом расширении

So1ovyov commented on ImageFeed/ServiceAccessLayer/OAuth2/OAuth2Service.swift:
```diff
@@ -0,0 +1,71 @@
+
+import Foundation
+
+final class OAuth2Service {
```
`OAuth2Service` — [синглтон](https://theswiftdev.com/swift-singleton-design-pattern/), то есть представляет глобальную точку доступа и запрещает создание других экземпляров
Это одно из условий ТЗ

So1ovyov commented on ImageFeed/ServiceAccessLayer/OAuth2/OAuthTokenResponseBody.swift:
```diff
@@ -0,0 +1,22 @@
+    let scope: String
+    let createdAt: Int
+    
+    private enum CodingKeys: String, CodingKey {
```
Для того, чтобы не работать с `snake_case` ключами, не описывать их в `CodingKeys`, ты можешь также 
настраивать `JSONDecoder` под тот вид данных, который приходит от сервера. 
Рекомендую изучить этот способ самостоятельно https://nemecek.be/blog/90/how-to-decode-snake-case-with-codable

So1ovyov commented on ImageFeed/ServiceAccessLayer/OAuth2/URLSession+data.swift:
```diff
@@ -0,0 +1,43 @@
+}
+
+extension URLSession {
+    func data(
```
Вот это расширение нужно применить в `OAuth2Service`

So1ovyov commented on ImageFeed/SingleImageView/SingleImageViewController.swift:
```diff
@@ -8,12 +8,10 @@
 
 final class SingleImageViewController: UIViewController {
-    var image: UIImage? {
+    var image: UIImage! {
```
Нарушение ТЗ, `force unwrap`

So1ovyov commented on ImageFeed/SplashViewController/SplashViewController.swift:
```diff
@@ -0,0 +1,98 @@
+    // MARK: - Private methods
+    
+    private func switchToTabBarController() {
+        DispatchQueue.main.async {
```
Нет необходимости оборачивать логику в асинхронный вызов.

So1ovyov commented on ImageFeed/SplashViewController/SplashViewController.swift:
```diff
@@ -0,0 +1,98 @@
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        
+        if let token = oauth2TokenStorage.token {
```
Предлагаю не создавать константу которая не используется, а просто проверить на `nil` токен
```
if oauth2TokenStorage.token != nil {
```

So1ovyov commented on ImageFeed/SplashViewController/SplashViewController.swift:
```diff
@@ -0,0 +1,98 @@
+extension SplashViewController: AuthViewControllerDelegate {
+    func authViewController(_ vc: AuthViewController, didAuthenticateWithCode code: String) {
+        dismiss(animated: true) { [weak self] in
+            guard let self = self else { return }
```
Подобные конструкции можно сделать немного короче:
```
guard let self else { return }
```

Fennec-meow committed:
Corrections:
1. Implicit unwrap and other force operations were removed from the project.
2. Constants were described inside the enum Constants.
3. Fixed the "WebView" controller.
4. If there is saved authorization data, the user is sent to the gallery flow, if not— to the authorization flow.
5. After previously successful authorization, the user enters the gallery immediately after "SplashViewController".
6. The completion block is called on the main thread (DispatchQueue.main).
7. A singleton has been added to the "OAuth2Service".
8. Erroneous cases were corrected when receiving optional values such as "URLComponents", "URL", "URLRequest".
9. Correction of the design in accordance with the design of the layout in Figma.

So1ovyov commented on ImageFeed/ServiceAccessLayer/OAuth2/OAuth2Service.swift:
```diff
@@ -0,0 +1,128 @@
+    
+    // MARK: - makeOAuthTokenRequest
+    
+    func makeOAuthTokenRequest(code: String) -> URLRequest {
```
Этот метод не используется в классе и содержит в себе логику которая повторяется в методе `makeOAuthTokenRequest` с 44 по 52

So1ovyov commented on ImageFeed/ServiceAccessLayer/OAuth2/OAuth2Service.swift:
```diff
@@ -0,0 +1,128 @@
+            + "&&code=\(code)"
+            + "&&grant_type=authorization_code",
+            relativeTo: baseURL                           // Опираемся на основной или базовый URL, которые содержат схему и имя хоста
+        )!
```
`Force unwrap` запрещен по условию ТЗ. 
В данном случае нужно безопасно извлечь значение например с помощью `guard`

So1ovyov commented on ImageFeed/ServiceAccessLayer/OAuth2/OAuth2Service.swift:
```diff
@@ -0,0 +1,128 @@
+    }
+}
+
+extension URLSession {
```
Это расширение просто существует в проекте и не используется в коде.

So1ovyov replied:
Я так и не понял зачем это расширение в проекте, если ты выше в методе `fetchOAuthToken` его не используешь и делаешь по сути тоже самое в коде метода. Предлагаю тогда все со 102 строки и ниже удалить, иначе мертвый код останется в проекте.

So1ovyov replied:
https://share.cleanshot.com/gpJjpthg

So1ovyov commented on ImageFeed/ServiceAccessLayer/OAuth2/OAuth2Service.swift:
```diff
@@ -0,0 +1,128 @@
+                    completion(.success(json.accessToken))
+                } catch {
+                    print("Decoding error: \(error)")
+                    completion(.failure(OAuthError.decodeError))
```
`Completion`-ы не вызваны на главном потоке.

So1ovyov commented on ImageFeed/ServiceAccessLayer/OAuth2/OAuth2Service.swift:
```diff
@@ -0,0 +1,128 @@
+    // MARK: - fetchOAuthToken
+    
+    func fetchOAuthToken(_ code: String, completion: @escaping (Result<String, Error>) -> Void) {
+        var components = URLComponents(string: "https://unsplash.com/oauth/token")!
```
Здесь также запрещенный `force unwrap` помог бы `guard`
```
guard var components = URLComponents(string: "https://unsplash.com/oauth/token") else { return }
```

So1ovyov commented on ImageFeed/SplashViewController/SplashViewController.swift:
```diff
@@ -0,0 +1,103 @@
+    // MARK: - Vars
+    
+    private let showAuthenticationScreenSegueIdentifier = "ShowAuthentication"
+    private let showTableSegueIdentifier = "ShowTable"
```
Эта константа не используется ее можно удалить.

So1ovyov commented on ImageFeed/SplashViewController/SplashViewController.swift:
```diff
@@ -0,0 +1,103 @@
+                self.switchToTabBarController()
+            case .failure:
+                
+                break
```
Лишняя пустая строка, можно удалить.

So1ovyov commented on ImageFeed/ServiceAccessLayer/OAuth2/OAuth2Service.swift:
```diff
@@ -0,0 +1,128 @@
+                    if response.statusCode < 200 || response.statusCode >= 300 {
+                        // Печатаем содержимое ответа для дополнительной информации
+                        if let data = data, let responseBody = String(data: data, encoding: .utf8) {
+                            print("Response Body: \(responseBody)")
```
Здесь не обработана ошибка не вызван `failure`

So1ovyov commented on ImageFeed/ServiceAccessLayer/OAuth2/OAuth2Service.swift:
```diff
@@ -0,0 +1,128 @@
+                        if let data = data, let responseBody = String(data: data, encoding: .utf8) {
+                            print("Response Body: \(responseBody)")
+                        }
+                        completion(.failure(OAuthError.codeError))
```
Здесь не распечатана ошибка

So1ovyov commented on ImageFeed/ServiceAccessLayer/OAuth2/OAuth2Service.swift:
```diff
@@ -0,0 +1,128 @@
+            URLQueryItem(name: "grant_type", value: "authorization_code")
+        ]
+        
+        if let url = components.url {
```
В этом методе и выше в случае ошибочного кейса, при получении опциональных значений, таких как URLComponents, URL, URLRequest нет `print`-ов, что необходимо по ТЗ.

Fennec-meow committed:
Corrections:
1. Implicit unwrap and other force operations were removed from the project.
2. Constants were described inside the enum Constants.
3. Fixed the "WebView" controller.
4. If there is saved authorization data, the user is sent to the gallery flow, if not— to the authorization flow.
5. After previously successful authorization, the user enters the gallery immediately after "SplashViewController".
6. The completion block is called on the main thread (DispatchQueue.main).
7. A singleton has been added to the "OAuth2Service".
8. Erroneous cases were corrected when receiving optional values such as "URLComponents", "URL", "URLRequest".
9. Correction of the design in accordance with the design of the layout in Figma.

So1ovyov commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,145 @@
+    
+    // MARK: - AuthorizeURL
+    
+    private let unsplashAuthorizeURLString = "https://unsplash.com/oauth/authorize"
```
Чтобы не хранить это внутри класса, можно перенести это в `enum Constants`, он как раз предназначен для хранения подобного

So1ovyov commented on ImageFeed/Profile/ProfileViewController.swift:
```diff
@@ -49,24 +59,42 @@ final class ProfileViewController: UIViewController {
-        button.tintColor = .ypRed
-        return button
+        // Безопасно получаем изображение
+        if let image = UIImage(named: "logout_button") {
```
Обрати внимание, у тебя много кода дублируется, можно решить вопрос немного проще:
```
    private lazy var logoutButton: UIButton = {
        let button = UIButton.systemButton(
            with: UIImage(named: "logout_button") ?? UIImage(),
            target: self,
            action: #selector(didTapLogoutButton)
        )
        button.translatesAutoresizingMaskIntoConstraints = false
        button.tintColor = .ypRed
        return button
    }()
```

So1ovyov commented on ImageFeed/ServiceAccessLayer/OAuth2/OAuth2Service.swift:
```diff
@@ -0,0 +1,131 @@
+        guard var components = URLComponents(string: "https://unsplash.com/oauth/token") else { return }
+        
+        components.queryItems = [
+            URLQueryItem(name: "client_id", value: Constants.accessKey),
```
При желание и эти строки для формирования `URLQueryItem` можно вынести в `enum Constants`

So1ovyov commented on ImageFeed/SplashViewController/SplashViewController.swift:
```diff
@@ -0,0 +1,115 @@
+    // MARK: - Vars
+    
+    private let showAuthenticationScreenSegueIdentifier = "ShowAuthentication"
+    //    private let showTableSegueIdentifier = "ShowWebView"
```
Закомментированный код можно удалить, если он не используется

So1ovyov commented on ImageFeed/SplashViewController/SplashViewController.swift:
```diff
@@ -0,0 +1,115 @@
+    
+    private func fetchOAuthToken(_ code: String) {
+        oauth2Service.fetchOAuthToken(code) { [weak self] result in
+            guard let self = self else { return }
```
По аналогии с методом выше:
```
guard let self else { return }
```

So1ovyov commented:
@Fennec-meow ты отлично потрудилась, критических замечаний нет, если будет время и возможность, предлагаю удалить лишнее в коде и применить несколько рекомендаций оставленных в ПР-е. 
Ты молодец, успехов в следующем спринте!
